File: scripts/julia/ex_lag_mult.jl
using LinearAlgebra
using DelimitedFiles

outdir = length(ARGS) >= 1 ? ARGS[1] : ""

A = [1.0, 1.0, 1.0]
E = [64.0, 64.0, 64.0]
L = [4.0, 4.0, 6.0]
alpha = [acos(3.0 / 4.0), -acos(3.0 / 4.0), 0.0]

Dvec = [
    1 2 5 6
    5 6 3 4
    1 2 3 4
]

N = maximum(Dvec)
I6 = Matrix{Float64}(I, N, N)
S1 = I6[Dvec[1, :], :]
S2 = I6[Dvec[2, :], :]
S3 = I6[Dvec[3, :], :]

kref = [
    1.0 0.0 -1.0 0.0
    0.0 0.0 0.0 0.0
    -1.0 0.0 1.0 0.0
    0.0 0.0 0.0 0.0
]

c = cos.(alpha)
s = sin.(alpha)

T1 = [
    c[1] s[1] 0.0 0.0
    -s[1] c[1] 0.0 0.0
    0.0 0.0 c[1] s[1]
    0.0 0.0 -s[1] c[1]
]
T2 = [
    c[2] s[2] 0.0 0.0
    -s[2] c[2] 0.0 0.0
    0.0 0.0 c[2] s[2]
    0.0 0.0 -s[2] c[2]
]
T3 = [
    c[3] s[3] 0.0 0.0
    -s[3] c[3] 0.0 0.0
    0.0 0.0 c[3] s[3]
    0.0 0.0 -s[3] c[3]
]

k1 = A[1] * E[1] / L[1] * kref
k2 = A[2] * E[2] / L[2] * kref
k3 = A[3] * E[3] / L[3] * kref

kg1 = transpose(T1) * k1 * T1
kg2 = transpose(T2) * k2 * T2
kg3 = transpose(T3) * k3 * T3

K = transpose(S1) * kg1 * S1 + transpose(S2) * kg2 * S2 + transpose(S3) * kg3 * S3

G = [
    1.0 0.0 0.0 0.0 0.0 0.0
    0.0 1.0 0.0 0.0 0.0 0.0
    0.0 0.0 -sin(deg2rad(60.0)) cos(deg2rad(60.0)) 0.0 0.0
]
Q = zeros(Float64, size(G, 1))
a_G = maximum(K)
Gbar = a_G * G
Qbar = a_G * Q

P = zeros(Float64, N)
P[6] = -10.0

AL = [K transpose(Gbar); Gbar zeros(Float64, size(G, 1), size(G, 1))]
solution = AL \ vcat(P, Qbar)

U = solution[1:N]
Lag = solution[(N + 1):end] * a_G
R = K[1:4, :] * U

ueg1 = U[Dvec[1, :]]
ueg2 = U[Dvec[2, :]]
ueg3 = U[Dvec[3, :]]

ue1 = T1 * ueg1
ue2 = T2 * ueg2
ue3 = T3 * ueg3

local_displacements = hcat(ue1, ue2, ue3)
member_forces = hcat(k1 * ue1, k2 * ue2, k3 * ue3)
constraint_residual = G * U - Q

if outdir != ""
    mkpath(outdir)
    writedlm(joinpath(outdir, "U.tsv"), reshape(U, :, 1), '\t')
    writedlm(joinpath(outdir, "Lag.tsv"), reshape(Lag, :, 1), '\t')
    writedlm(joinpath(outdir, "R.tsv"), reshape(R, :, 1), '\t')
    writedlm(joinpath(outdir, "member_forces.tsv"), member_forces, '\t')
    writedlm(joinpath(outdir, "local_displacements.tsv"), local_displacements, '\t')
    writedlm(joinpath(outdir, "constraint_residual.tsv"), reshape(constraint_residual, :, 1), '\t')
else
    println("U = ", U)
    println("Lag = ", Lag)
    println("R = ", R)
    println("local_displacements = ", local_displacements)
    println("member_forces = ", member_forces)
end

File: scripts/matlab/ex_lag_mult.m
function result = ex_lag_mult(output_dir)
if nargin < 1
    output_dir = "";
end

A = [1 1 1];
E = [64 64 64];
L = [4 4 6];
alpha = [acos(3 / 4), -acos(3 / 4), 0];

Dvec = [
    1 2 5 6
    5 6 3 4
    1 2 3 4
];

N = max(Dvec, [], "all");
[Nelem, dof2] = size(Dvec);
dof = dof2 / 2;
K = zeros(N, N);

for e = 1:Nelem
    [k, T] = k_truss(A(e), E(e), L(e), alpha(e));
    kg = T' * k * T;
    for r = 1:(2 * dof)
        m = Dvec(e, r);
        if m ~= 0
            for s = 1:(2 * dof)
                n = Dvec(e, s);
                if n ~= 0
                    K(m, n) = K(m, n) + kg(r, s);
                end
            end
        end
    end
end

G = [
    1 0 0 0 0 0
    0 1 0 0 0 0
    0 0 -sind(60) cosd(60) 0 0
];
Q = zeros(size(G, 1), 1);
a_G = max(K, [], "all");
Gbar = a_G * G;
Qbar = a_G * Q;

P = zeros(N, 1);
P(6) = -10;

AL = [K, Gbar'; Gbar, zeros(size(G, 1), size(G, 1))];
BL = [P; Qbar];
solution = AL \ BL;

Lag = solution((N + 1):end) * a_G;
U = solution(1:N);
R = K(1:4, :) * U;

member_forces = zeros(4, Nelem);
for e = 1:Nelem
    [k, T] = k_truss(A(e), E(e), L(e), alpha(e));
    ueg = U(Dvec(e, :));
    ue = T * ueg;
    member_forces(:, e) = k * ue;
end

constraint_residual = G * U - Q;

result = struct( ...
    "U", U, ...
    "Lag", Lag, ...
    "R", R, ...
    "member_forces", member_forces, ...
    "constraint_residual", constraint_residual ...
);

if strlength(output_dir) > 0
    if ~isfolder(output_dir)
        mkdir(output_dir);
    end
    write_tsv(fullfile(output_dir, "U.tsv"), U);
    write_tsv(fullfile(output_dir, "Lag.tsv"), Lag);
    write_tsv(fullfile(output_dir, "R.tsv"), R);
    write_tsv(fullfile(output_dir, "member_forces.tsv"), member_forces);
    write_tsv(fullfile(output_dir, "constraint_residual.tsv"), constraint_residual);
elseif nargout == 0
    disp(result);
end
end

function [k, T] = k_truss(A, E, L, alpha)
c = cos(alpha);
s = sin(alpha);
T = [
    c s 0 0
    -s c 0 0
    0 0 c s
    0 0 -s c
];
k = A * E / L * [
    1 0 -1 0
    0 0 0 0
    -1 0 1 0
    0 0 0 0
];
end

function write_tsv(path, data)
writematrix(data, path, "Delimiter", "tab", "FileType", "text");
end
